metaflow/package/mfenv.py

```diff
@@ -0,0 +1,867 @@
+
+        return filter(importable_name, opt_names)
+
+    packages_distributions = _packages_distributions
+
+
+if TYPE_CHECKING:
+    import pathlib
+
+
+_cached_distributions = None
+
+name_normalizer = re.compile(r"[-_.]+")
+
+
+class AddToPackageType(IntEnum):
+    CODE_ONLY = 0x100
+    CONFIG_ONLY = 0x200
+    FILES_ONLY = 0x400
+    ALL = 0xFFF
+    CODE_FILE = 0x501
+    CODE_MODULE = 0x502
+    CODE_METAFLOW = 0x504
+    CONFIG_FILE = 0x601
+    CONFIG_CONTENT = 0x202
+
+
+def modules_to_distributions() -> Dict[str, List[metadata.Distribution]]:
+    """
+    Return a mapping of top-level modules to their distributions.
+
+    Returns
+    -------
+    Dict[str, List[metadata.Distribution]]
+        A mapping of top-level modules to their distributions.
+    """
+    global _cached_distributions
+    if _cached_distributions is None:
+        _cached_distributions = {
+            k: [metadata.distribution(d) for d in v]
+            for k, v in packages_distributions().items()
+        }
+    return _cached_distributions
+
+
+_ModuleInfo = NamedTuple(
+    "_ModuleInfo", [("name", str), ("root_paths", Set[str]), ("module", ModuleType)]
+)
+
+
+class PackagedDistribution(metadata.Distribution):
+    """
+    A Python Package packaged within a MFEnv. This allows users to use use importlib
+    as they would regularly and the packaged Python Package would be considered as a
+    distribution even if it really isn't (since it is just included in the PythonPath).
+    """
+
+    def __init__(self, root: str, content: Dict[str, str]):
+        self._root = Path(root)
+        self._content = content
+
+    # Strongly inspired from PathDistribution in metadata.py
+    def read_text(self, filename: Union[str, os.PathLike]) -> Optional[str]:
+        if str(filename) in self._content:
+            return self._content[str(filename)]
+        return None
+
+    read_text.__doc__ = metadata.Distribution.read_text.__doc__
+
+    # Returns a metadata.SimplePath but not always present in importlib.metadata libs so
+    # skipping return type.
+    def locate_file(self, path: Union[str, os.PathLike]):
+        return self._root / path
+
+
+class PackagedDistributionFinder(metadata.DistributionFinder):
+
+    def __init__(self, dist_info: Dict[str, Dict[str, str]]):
+        self._dist_info = dist_info
+
+    def find_distributions(self, context=metadata.DistributionFinder.Context()):
+        if context.name is None:
+            # Yields all known distributions
+            for name, info in self._dist_info.items():
+                yield PackagedDistribution(
+                    os.path.join(get_metaflow_root(), name), info
+                )
+        name = name_normalizer.sub("-", context.name).lower()
+        if name in self._dist_info:
+            yield PackagedDistribution(
+                os.path.join(get_metaflow_root(), context.name),
+                self._dist_info[name],
+            )
+        return None
+
+
+class MFEnv:
```

Comment:
can MetaflowPackage instead be the way that rest of the code accesses functionality here? will make it easier for all the decorator implementations which already have access to the package object to get access to info files etc.

Reply:
reading through the code, MFEnv seems like an internal implementation detail for MetaflowPackage. If we can keep it as such throughout the code, that will be great!